utils/ensemble_tools.py

```diff
@@ -0,0 +1,158 @@
+
+    print(
+        "{0} parameters found, requiring {1} divisions per parameter. "
+        "Creating {2} configurations...".format(
+            len(parameter_dict.keys()), n_dirs, n_dirs ** len(parameter_dict.keys())
+        )
+    )
+
+    # Calculate all of the parameter values and put them in a dict
+    scan_points = {}
+    list_all = []
+    for parameter in parameter_dict.keys():
+        param_value = parameter_dict[parameter][0]
+        scan_points[parameter] = []
+        for i in range(n_dirs):
+            scan_points[parameter].append(param_value)
+            list_all.append("{0}_{1}".format(parameter, i))
+            param_value += (
+                parameter_dict[parameter][1] - parameter_dict[parameter][0]
+            ) / float(n_dirs - 1)
+
+    # Create the combinations of these parameters.
+    combinations = list(itertools.combinations(list_all, len(parameter_dict.keys())))
+    all_configs = []
+    for comb in combinations:
+        record_comb = True
+        for i in range(len(comb) - 1):
+            search_term = comb[i][: len(comb[i]) - comb[i][::-1].find("_")]
+            for j in range(i + 1, len(comb)):
+                if comb[j].startswith(search_term):
+                    record_comb = False
+                    break
+        if record_comb:
+            all_configs.append(comb)
+
+    # If destructive, let's delete the whole tree if it already exists
+    if destructive and os.path.isdir(sweep_path):
+        shutil.rmtree(sweep_path)
+    # Check whether the copy directory exists
+    copy_files = os.path.isdir(copy_dir)
+    # Now loop over each combination, create a directory, copy the
+    # configs and conditions, and edit the configuation to contain
+    # the correct parameters.
+    for comb in all_configs:
+        dirName = "-".join(comb)
+        dir_path = "{0}/SWEEP/{1}".format(sweep_path, dirName)
+        make_directory(dir_path, destructive)
+        if copy_files:
+            copy_dir_contents(dir_path, copy_dir)
+        # Create parameter map for conditions encoding
+        parameters = {}
+        for parameter in parameter_dict.keys():
+            param_index = [item for item in comb if item.startswith(parameter)][
+                0
+            ].split("_")[-1]
+            parameters[parameter] = scan_points[parameter][int(param_index)]
+        if edit_file:
+            encode_conditions_file(dir_path + "/" + edit_file, parameters)
+    return 0
+
+
+def copy_dir_contents(dir_path, copy_dir):
+    """Copy the contents of copy_dir -> dir_path"""
+
+    for f in os.listdir(copy_dir):
+        shutil.copy(copy_dir + f, dir_path)
+
+
+def make_directory(directory_name, destructive):
+    """Make the directory tree at directory_name."""
+
+    if os.path.isdir(directory_name):
+        if destructive:
+            shutil.rmtree(directory_name)
+        else:
+            print("Path already exists and " "not in destructive mode! Returning")
+            return 0
```

Comment:
It would be more idiomatic to raise an exception here for example
```Python
msg = "Path already exists and not in destructive mode."
raise FileExistsError(msg)
```
and then when calling the function just assume it runs correctly unless an exception is raised which will cause script to exit unless explicitly handled with a `try... except`.